rp2040-hal/src/pwm/mod.rs

```diff
@@ -0,0 +1,581 @@
+
+    ///
+    fn output_to<
+        G: PinId + BankPinId + ValidPwmOutputPin<Self::S, Self::C>,
+        M: PinMode + ValidPinMode<G>,
+    >(
+        &mut self,
+        pin: Pin<G, M>,
+    );
+}
+
+/// Marks this Channel as having input (Only channels B)
+pub trait PwmInputChannel: PwmOutputChannel {
+    ///
+    fn input_from<G: PinId + BankPinId + ValidPwmInputPin<Self::S>, M: PinMode + ValidPinMode<G>>(
+        &mut self,
+        pin: Pin<G, M>,
+        mode: u8,
+    );
+}
+
+/// Make sure we can't free an GPIO pin while still keeping it attached to pwm
+/// TODO: Maybe FunctionPWM should be private?
+pub trait NonPwmPinMode: Sealed {}
+
+impl NonPwmPinMode for FunctionClock {}
+impl NonPwmPinMode for FunctionI2C {}
+impl NonPwmPinMode for FunctionPio0 {}
+impl NonPwmPinMode for FunctionPio1 {}
+impl NonPwmPinMode for FunctionSpi {}
+impl NonPwmPinMode for FunctionUart {}
+impl NonPwmPinMode for FunctionUsbAux {}
+impl NonPwmPinMode for FunctionXip {}
+impl<C: InputConfig> NonPwmPinMode for Input<C> {}
+impl<C: OutputConfig> NonPwmPinMode for Output<C> {}
+
+///
+pub struct PwmPinToken<G: PinId + BankPinId> {
+    pin: Pin<G, FunctionPwm>,
+}
+///
+impl<G: PinId + BankPinId> PwmPinToken<G> {
+    ///
+    pub fn into_mode<N: PinMode + ValidPinMode<G> + NonPwmPinMode>(self) -> Pin<G, N> {
+        self.pin.into_mode::<N>()
+    }
+}
+
+impl Slices {
+    /// Free the pwm registers from the pwm hal struct while consuming it.
+    pub fn free(self) -> PWM {
+        self._pwm
+    }
+
+    //     /// Enable multiple slices at the same time to make their counters sync up.
+    //     ///
+    //     /// You still need to call `slice` to get an actual slice
+    //     pub fn enable_simultanous<S: SliceId>(&mut self, bits: u8) {
+    //         // Enable all slices at the same time
+    //         unsafe {
+    //             &(*pac::PWM::ptr())
+    //                 .en
+    //                 .modify(|r, w| w.bits(((r.bits() as u8) | bits) as u32));
+    //         }
+    //     }
+}
+
+impl<S: SliceId, M: SliceMode + ValidSliceMode<S>> Slice<S, M> {
+    /// Set a default config for the slice
+    pub fn default_config(&mut self) {
+        self.regs.write_ph_correct(false);
+        self.regs.write_div_int(1); // No divisor
+        self.regs.write_div_frac(0); // No divisor
+        self.regs.write_inv_a(false); //Don't invert the channel
+        self.regs.write_inv_b(false); //Don't invert the channel
+        self.regs.write_top(0xffffu16); // Wrap at max
+        self.regs.write_ctr(0x0000u16); //Reset the counter
+        self.regs.write_cc_a(0); //Default duty cycle of 0%
+        self.regs.write_cc_b(0); //Default duty cycle of 0%
+    }
+
+    /// Enable Slice
+    #[inline]
+    pub fn enable(&mut self) {
+        self.regs.enable(true);
+    }
+
+    /// Disable slice
+    #[inline]
+    pub fn disable(&mut self) {
+        self.regs.enable(true)
+    }
+
+    #[doc = "Enables phase correct mode"]
```

Comment:
Why using `#[doc = ""]` and not `///`? Also
```suggestion
    #[doc = "Enable phase correct mode"]
```